File: EmojiPanel/src/main/java/com/binlee/emoji/helper/LogHelper.java
package com.binlee.emoji.helper;

import android.os.SystemClock;
import android.util.Log;

import com.binlee.emoji.BuildConfig;

/**
 * Created on 19-1-18.
 *
 * @author leebin
 */
public final class LogHelper {

    private static final boolean LOGGABLE = BuildConfig.DEBUG;
    private static final int MAX_LOG_LENGTH = 4000;
    private static final double MILLIS_MULTIPLIER = 1d / Math.pow(10, 6);

    private LogHelper() {
        throw new AssertionError("no instance");
    }

    public static double elapsedMillis(long logTime) {
        return (logTime() - logTime) * MILLIS_MULTIPLIER;
    }

    public static long logTime() {
        return SystemClock.elapsedRealtimeNanos();
    }

    public static void debug(String tag, String msg) {
        debug(tag, msg, null);
    }

    /**
     * copy from OkHttp
     */
    public static void debug(String tag, String msg, Throwable thr) {
        if (!LOGGABLE) {
            return;
        }
        if (thr != null) {
            msg = msg + '\n' + Log.getStackTraceString(thr);
        }

        // Split by line, then ensure each line can fit into Log's maximum length.
        for (int i = 0, length = msg.length(); i < length; i++) {
            int newline = msg.indexOf('\n', i);
            newline = newline != -1 ? newline : length;
            do {
                int end = Math.min(newline, i + MAX_LOG_LENGTH);
                Log.println(Log.DEBUG, tag, msg.substring(i, end));
                i = end;
            } while (i < newline);
        }
    }

    public static void dumpStackTrace(String tag) {
        final StackTraceElement[] traceElements = Thread.currentThread().getStackTrace();
        for (final StackTraceElement element : traceElements) {
            Log.e(tag, "" + element);
        }
        Log.i(tag, "<============= the current stack trace ==============>");
    }
}

File: EmojiPanel/src/main/java/com/binlee/emoji/helper/LogHelper.kt
package com.binlee.emoji.helper

import android.os.SystemClock
import android.util.Log
import com.binlee.emoji.BuildConfig
import kotlin.math.min
import kotlin.math.pow

/**
 * Created on 19-1-18.
 *
 * @author leebin
 */
object LogHelper {

    private val LOGGABLE = BuildConfig.DEBUG
    private const val MAX_LOG_LENGTH = 4000
    private val MILLIS_MULTIPLIER = 1.0 / 10.0.pow(6.0)

    fun elapsedMillis(logTime: Long): Double {
        return (logTime() - logTime) * MILLIS_MULTIPLIER
    }

    fun logTime(): Long {
        return SystemClock.elapsedRealtimeNanos()
    }

    @JvmStatic
    fun debug(tag: String, msg: String) {
        debug(tag, msg, thr = null)
    }

    /**
     * copy from OkHttp
     */
    @JvmStatic
    fun debug(tag: String?, msg: String, thr: Throwable? = null) {
        var input = msg
        if (!LOGGABLE) {
            return
        }
        if (thr != null) {
            input = "$msg${Log.getStackTraceString(thr)}"
        }

        // Split by line, then ensure each line can fit into Log's maximum length.
        var i = 0
        val length = input.length
        while (i < length) {
            var newline = input.indexOf('\n', i)
            newline = if (newline != -1) newline else length
            do {
                val end = min(newline, i + MAX_LOG_LENGTH)
                Log.println(Log.DEBUG, tag, input.substring(i, end))
                i = end
            } while (i < newline)
            i++
        }
    }

    fun dumpStackTrace(tag: String?) {
        for (element in Thread.currentThread().stackTrace) {
            Log.e(tag, "" + element)
        }
        Log.i(tag, "<============= the current stack trace ==============>")
    }
}